aineko/extras/connectors/http_poller.py

```diff
@@ -0,0 +1,257 @@
+        HTTPoller:
+          class: aineko.extras.HTTPPoller
+          outputs:
+            - test_http
+          node_params:
+            url: "https://example.com"
+            headers:
+              auth: "Bearer {$SECRET_NAME}"
+            data: {"Greeting": "Hello, world!"}
+    ```
+
+    Secrets can be injected (from environment) into the `url`, `headers`, and
+    `data` fields by passing a string with the following format:
+    `{$SECRET_NAME}`. For example, if you have an environment variable named
+    `SECRET_NAME`that contains the value `SECRET_VALUE`, you can inject it into
+    the url field by passing `https://example.com?secret={$SECRET_NAME}` as the
+    url. The connector will then replace `{$SECRET_NAME}` with `SECRET_VALUE`
+    before connecting to the HTTP endpoint.
+
+    Note that the `outputs` field is required and must contain exactly one
+    output dataset. The output dataset will contain the data returned by the
+    endpoint.
+
+    By default, this node will poll the endpoint every 5 seconds and timeout
+    after 10 seconds. If the request fails, it will retry every 5 seconds
+    forever. Status codes in the 200s are considered success codes and no
+    headers, data, auth, params, or json will be attached to the request.
+    """
+
+    # Poll settings
+    last_poll_time = time.time()
+    retry_count = 0
+
+    def _pre_loop_hook(self, params: Optional[Dict] = None) -> None:
+        """Initializes connection to API."""
+        try:
+            if params is not None:
+                self.http_poller_params = ParamsHTTPPoller(**params)
+            else:
+                raise ValueError(
+                    "No params provided to HTTPPoller connector. "
+                    "Node requires at least a url param."
+                )
+        except Exception as err:  # pylint: disable=broad-except
+            # Cast pydantic validation error to ValueError
+            # so that it can be properly caught by the node
+            # Note: this is required because pydantic errors
+            # are not pickleable
+            raise ValueError(
+                "Failed to cast params to ParamsHTTPPoller type. "
+                f"The following error occurred: {err}"
+            ) from err
+
+        # Ensure only one output dataset is provided
+        output_datasets = [
+            dataset for dataset in self.producers.keys() if dataset != "logging"
+        ]
+        if len(output_datasets) > 1:
+            raise ValueError(
+                "Only one output dataset is allowed for the "
+                "HTTPoller connector. "
+                f"{len(output_datasets)} datasets given."
+            )
+        self.output_dataset = output_datasets[0]
+
+        # Create a session
+        self.log(
+            f"Creating new session to endpoint {self.http_poller_params.url}."
+        )
+        self.session = requests.Session()
+
+    def _execute(self, params: Optional[Dict] = None) -> None:
```

Comment:
Time permitting, I would break this function into smaller functions based on `polling` and `parse_and_write`. It's quite a bit longer than the ideal function length of ~40 lines. https://google.github.io/styleguide/pyguide.html#318-function-length

Reply:
fixed